runtime/src/core/memory.js

```diff
@@ -1,26 +1,817 @@
+      throw 'Out of memory';
+    }
+    trace('\tpopulateHeader')
+    this.populateHeader(rawPtr);
+    trace('\tend_populateHeader')
+    let userPtr = rawPtr + this._headerSize;
+    trace(`malloc: ${this._memdump(userPtr)}`);
+    if (TRACE_MEMORY) {
+      this._allocatedAddresses.add(userPtr);
+      if (typeof this._maxAddress === 'undefined') {
+        this._maxAddress = rawPtr + size + this._headerSize;
+      } else {
+        this._maxAddress = Math.max(rawPtr + size + this._headerSize, this._maxAddress);
+      }
+      if (typeof this._minAddress === 'undefined') {
+        this._minAddress = rawPtr;
+      } else {
+        this._minAddress = Math.min(rawPtr, this._minAddress);
+      }
+      this._timesAllocated[userPtr] = (this._timesAllocated[userPtr] || 0) + 1;
+      this._freedAddresses.delete(userPtr);
+    }
+    this._markIncRefSource(userPtr, 'MALLOC');
+    this._colors[userPtr] = GREEN;
+    return userPtr; // offset by headerSize
+  }
+
+  populateHeader(rawPtr) {
+    let tag = 0x0; // reserved
+    trace('populateHeader');
+    var heap = new Uint8Array(this._memory.buffer);
+    for (let i = 0; i < this._headerSize; ++i) {
+      heap[rawPtr + i] = 0;
+    }
+    this._setRefCount(rawPtr, 1);
+    //heap[rawPtr + 2] = 1; // <- init refCount at 1 (from this allocation)
+    heap[rawPtr + 3] = tag & 0b1111; // <- 4-bit tag
+  }
+
+  // [TODO] These next three methods can probably be made more efficient
+  _getRefCount(userPtr) {
+    trace('_getRefCount');
+    let rawPtr = (userPtr & (~7)) - this._headerSize;
+    //let rawPtr = userPtr - 3;
+    let heap = new Uint8Array(this._memory.buffer);
+    trace(`\tas binary: ${this._toBinary(heap[rawPtr], 8, 8)}|${this._toBinary(heap[rawPtr + 1], 8, 8)}|${this._toBinary(heap[rawPtr + 2], 8, 8)}`)
+    let count = heap[rawPtr] & 0b0111; // reserved bit should always be zero, but let's be safe
+    count = count << 8;
+    count = count | heap[rawPtr + 1];
+    count = count << 8;
+    count = count | heap[rawPtr + 2];
+    trace(`\t${count}`);
+    return count;
+  }
+
+  _getValueTag(userPtr) {
+    trace('_getValueTag');
+    let rawPtr = userPtr - this._headerSize;
+    let heap = new Uint8Array(this._memory.buffer);
+    return heap[rawPtr + 3];
+  }
+
+  _toHex(n, minWidth) {
+    if (!TRACE_MEMORY) {
+      return;
+    }
+    return toHex(n, minWidth);
+  }
+
+  _toBinary(n, minWidth, maxWidth) {
+    if (!TRACE_MEMORY) {
+      return;
+    }
+    let ret = toBinary(n, minWidth);
+    if (maxWidth && ret.length > maxWidth) {
+      if (ret.substring(0, ret.length - maxWidth).includes('1')) {
+        console.warn('chopping off at least one 1!');
+      }
+      ret = ret.substring(ret.length - maxWidth);
+    }
+    return ret;
+  }
+
+  _memdump(userPtr) {
+    let heap = new Uint8Array(this._memory.buffer);
+    if (!TRACE_MEMORY) {
+      return undefined; // improve performance; this is a very hot path in production
+    }
```

Comment:
Terser will convert this function into a no-op and then should even remove the call to it.

```suggestion
    if (!TRACE_MEMORY) {
      return;
    }
    let heap = new Uint8Array(this._memory.buffer);
```